src/Features/Core/Portable/IntroduceVariable/AbstractIntroduceParameterService.cs

```diff
@@ -0,0 +1,829 @@
+                    parameterName, editor, allOccurrences, cancellationToken).ConfigureAwait(false);
+                var semanticModel = await originalDocument.GetRequiredSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+                var parameterType = semanticModel.GetTypeInfo(expressionCopy, cancellationToken).Type ??
+                    semanticModel.Compilation.ObjectType;
+                var refKind = syntaxFacts.GetRefKindOfArgument(expressionCopy);
+                var parameter = generator.ParameterDeclaration(name: parameterName,
+                    type: generator.TypeExpression(parameterType), refKind: refKind);
+                editor.InsertParameter(containingMethod, insertionIndex, parameter);
+            }
+
+            return modifiedSolution.WithDocumentSyntaxRoot(originalDocument.Id, editor.GetChangedRoot());
+        }
+
+        /// <summary>
+        /// Goes through all of the invocations and replaces the expression with identifiers from the invocation 
+        /// arguments and rewrites the call site with the updated expression as a new argument.
+        /// </summary>
+        private async Task<Solution> RewriteSolutionAsync(Document originalDocument,
+            TExpressionSyntax expression, IMethodSymbol methodSymbol, SyntaxNode containingMethod,
+            ImmutableDictionary<Document, ImmutableArray<TInvocationExpressionSyntax>> callSites,
+            bool allOccurrences, string parameterName, CancellationToken cancellationToken)
+        {
+            // Need a copy of the expression to use to find the original expression
+            // because it gets modified when tracking the nodes.
+            var expressionCopy = expression;
+            var mappingDictionary = await MapExpressionToParametersAsync(originalDocument, expression, cancellationToken).ConfigureAwait(false);
+
+            // Need to track the nodes in the expression so that they can be found later.
+            expression = expression.TrackNodes(mappingDictionary.Keys);
+            var identifiers = expression.DescendantNodes().Where(node => node is TIdentifierNameSyntax);
+
+            var modifiedSolution = originalDocument.Project.Solution;
+
+            foreach (var (project, projectCallSites) in callSites.GroupBy(kvp => kvp.Key.Project))
+            {
+                var compilation = await project.GetRequiredCompilationAsync(cancellationToken).ConfigureAwait(false);
+                foreach (var (document, invocations) in projectCallSites)
+                {
+                    modifiedSolution = await ModifyDocumentInvocationsAndIntroduceParameterAsync(compilation,
+                        originalDocument, document, modifiedSolution, methodSymbol, containingMethod, expressionCopy,
+                        expression, allOccurrences, parameterName, invocations, cancellationToken).ConfigureAwait(false);
+                }
+            }
+
+            return modifiedSolution;
+        }
```

Comment:
this method looks nearly identical to RewriteSolutionWithNewMethodAsync

i woudl expect one top level RewriteSolution, which processes all documents unformly.  then you have code to handle both the non-originating docs, and the originating doc.  And the code for handling either of those, does the unique logic based on 'trampoline/overload'

Reply:
abstracted a bit